### PR TITLE
Fixes bugs in the Cohere Blueprint

### LINKED ISSUE #1344: [BUG] Cohere Blueprint issue
**What is the bug?**
When ingesting data using Cohere's blueprint and neural search the ingestion pipeline returns: 
```json
{
  "took": 0,
  "ingest_took": 2,
  "errors": true,
  "items": [
    {
      "create": {
        "_index": "cohere-index",
        "_id": "1",
        "status": 400,
        "error": {
          "type": "illegal_argument_exception",
          "reason": "Must provide pre_process_function for predict action to process text docs input."
        }
      }
    },
  ]
}
```

**How can one reproduce the bug?**
DevTools reproduction: 
```
# Create Connector: 
POST /_plugins/_ml/connectors/_create
{
   "name": "Cohere Connector",
   "description": "External connector for connections into Cohere",
   "version": "1.0",
   "protocol": "http",
   "credential": {
           "cohere_key": "<INSERT KEY>"
       },
    "parameters": {
      "model": "embed-english-v2.0",
      "truncate": "END"
    },
   "actions": [{
       "action_type": "predict",
       "method": "POST",
       "url": "https://api.cohere.ai/v1/embed",
       "headers": {
               "Authorization": "Bearer ${credential.cohere_key}"
           },
       "request_body": "{ \"texts\": ${parameters.texts}, \"truncate\": \"${parameters.truncate}\", \"model\": \"${parameters.model}\" }"
       }]
}


# Create a model related to the cluster
POST /_plugins/_ml/models/_register
{
    "name": "embed-english-v2.0",
    "function_name": "remote",
    "description": "test model",
    "connector_id": "9vxqm4oBYFYMFuLi9oOR"
}

# Deploy the model
POST /_plugins/_ml/models/_Pxrm4oBYFYMFuLiG4PH/_deploy

PUT _ingest/pipeline/cohere-ingest-pipeline
{
  "description": "Cohere Neural Search Pipeline",
  "processors" : [
    {
      "text_embedding": {
        "model_id": "_Pxrm4oBYFYMFuLiG4PH",
        "field_map": {
          "content": "content_embedding"
        }
      }
    }
  ]
}


# Create KNN index. Note* need to match space to model space. eg embed-english-v2.0 recommends cosine similarity
PUT /cohere-index
{
	"settings": {
		"index.knn": true,
		"default_pipeline": "cohere-ingest-pipeline"
	},
	"mappings": {
		"properties": {
			"content_embedding": {
				"type": "knn_vector",
				"dimension": 4096,
				"method": {
					"name": "hnsw",
					"space_type": "cosinesimil",
					"engine": "nmslib"
				}
			},
			"content": {
				"type": "text"
			}
		}
	}
}


POST /_bulk
{ "create" : { "_index" : "cohere-index", "_id" : "1" } }
{ "content":"Testing neural search"}
{ "create" : { "_index" : "cohere-index", "_id" : "2" } }
{ "content":"What are we doing"}
{ "create" : { "_index" : "cohere-index", "_id" : "3" } }
{ "content":"This should exist"}


GET /cohere-index/_search
{
  "query": {
    "bool" : {
      "should" : [
        {
          "script_score": {
              "neural": {
                "content_embedding": {
                  "query_text": "How do I ingest to opensearch",
                  "k": 10
              }
            },
            "script": {
              "source": "_score * 1.5"
            }
          }
        }
        ,
        {
          "script_score": {
            "query": {
              "match": { "content": "I want information about the new compression algorithems in OpenSearch" }
            },
            "script": {
              "source": "_score * 1.7"
            }
          }
        }
      ]
    }
  }
}
```

**What is the expected behavior?**
Expect data to be ingested. 

**What is your host/environment?**
 - OpenSearch: 2.9

### PR DESCRIPTION
**Pull request by dtaivpp**
### Description
1. Fixes Cohere prompt template from $parameters.texts to $parameters.prompt
2. Adds in the pre/post steps needed for Cohere pipeline. 
 
### Issues Resolved
Resolves #1344
 
### Check List
- [N/A] New functionality includes testing.
  - [N/A] All tests pass
- [N/A] New functionality has been documented.
  - [N/A] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md [hunk 1]
```diff
@@ -24,7 +24,9 @@ POST /_plugins/_ml/connectors/_create
       "headers": {
         "Authorization": "Bearer ${credential.cohere_key}"
       },
-      "request_body": "{ \"texts\": ${parameters.texts}, \"truncate\": \"${parameters.truncate}\", \"model\": \"${parameters.model}\" }"
+      "request_body": "{ \"texts\": ${parameters.prompt}, \"truncate\": \"${parameters.truncate}\", \"model\": \"${parameters.model}\" }", 
+      "pre_process_function": "connector.pre_process.cohere.embedding",
+			"post_process_function": "connector.post_process.cohere.embedding"
```
Comment (dhrubo-os) [inline docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:29]: Can we fix the indentation here?

File: docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md [hunk 2]
```diff
@@ -24,7 +24,9 @@ POST /_plugins/_ml/connectors/_create
       "headers": {
         "Authorization": "Bearer ${credential.cohere_key}"
       },
-      "request_body": "{ \"texts\": ${parameters.texts}, \"truncate\": \"${parameters.truncate}\", \"model\": \"${parameters.model}\" }"
+      "request_body": "{ \"texts\": ${parameters.prompt}, \"truncate\": \"${parameters.truncate}\", \"model\": \"${parameters.model}\" }",
```
Comment (dhrubo-os) [inline docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:27]: Cohere expects `texts`, please check [this doc](https://docs.cohere.com/reference/embed)

File: docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md (hunk 1 above)
Comment (dhrubo-os) [reply docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:29]: @ylwu-amzn should we add these two fields in all the connector blueprints?

Comment (ylwu-amzn) [reply docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:29]: Not all, these two functions should be added on demand. For text embedding model, they are usually used for neural search, so they need these two functions.

File: docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md (hunk 2 above)
Comment (ylwu-amzn) [reply docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:27]: Actually this is a java code issue https://github.com/opensearch-project/ml-commons/blob/2.x/common/src/main/java/org/opensearch/ml/common/connector/MLPreProcessFunction.java#L31

We should fix the java code in next release, but for now let's fix the doc first. When java code fixed, we should change this back to `texts`

Comment (dhrubo-os) [reply docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:27]: But that's only if we use `pre_process_function`, right?

Comment (ylwu-amzn) [reply docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:27]: Yes

**Review by ylwu-amzn — DISMISSED** [review]

**Comment by dtaivpp** [comment]
@dhrubo-os you mind taking a look one more time? We got the indentation fixed and I think we have landed on that the primary way the connector is getting used is with text which needs the pre-processing/post-processing steps.

**Comment by dtaivpp** [comment]
Would it cause issues if these actions were included and the model wasn't used in neural search? By this I mean does their presence cause any issues for other use cases?

**Review by saratvemulapalli — APPROVED** [review]

**Review by dhrubo-os — APPROVED** [review]

**Comment by ylwu-amzn** [comment]
> Would it cause issues if these actions were included and the model wasn't used in neural search? By this I mean does their presence cause any issues for other use cases?

By adding these functions the model can still run successfully even not used in neural search. But the output format will be different. So if not for neural search, it's up to user to add these functions or not. If user prefer to use the default response from Cohere, it's ok to remove the function.

File: docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md (hunk 2 above)
Comment (ylwu-amzn) [inline docs/remote_inference_blueprints/cohere_connector_embedding_blueprint.md:27]: @dtaivpp This PR already fixed the input filed name https://github.com/opensearch-project/ml-commons/pull/1274, We don't need to change `parameters.texts` to `parameters.prompt` with that PR

Can you rebase and update this PR?

**Review by ylwu-amzn — APPROVED** [review]

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**